### DISCUSSION TITLE
First person shooter weapon in viewspace?

### DISCUSSION THREAD

**Discussion by ssansovich** [raysan5/raylib#3130] [Q&A]
Hi everyone - I'm trying to build a simple 3D first person shooter. I'm having a very hard time figuring out how to render the weapon in an FPS game relative to the camera.

My current approach is to try to get the weapon to render in view space (rather than world space) using a custom shader but for the life of me I can't get it to render correctly. It seems like the object is just warped and sheared (though pixels are getting drawn to the screen).

Any suggestions for how to fix this or to do it better?

Thanks!

```
void draw_weapon(Model model, Camera *camera, float offset_x, float offset_y, float offset_z, float rot_x, float rot_y, float rot_z) {
    // Model
    Vector3 position = { offset_x, offset_y, offset_z };
    Matrix mat_tran = MatrixTranslate(position.x, position.y, position.z);

    Matrix matRotX = MatrixRotateX(rot_x);
    Matrix matRotY = MatrixRotateY(rot_y);
    Matrix matRotZ = MatrixRotateZ(rot_z);
    Matrix mat_rot = MatrixMultiply(MatrixMultiply(matRotZ, matRotY), matRotX);

    Matrix model_matrix = MatrixMultiply(mat_tran, mat_rot);

    // View
    Matrix view = MatrixIdentity();

    // Projection
    Matrix projection = GetCameraProjectionMatrix(camera, 1600.0f/900.0f);

    // Compute the model-view-projection matrix
    Matrix mvp = MatrixMultiply(projection, MatrixMultiply(view, model_matrix));

    int mvpLoc = GetShaderLocation(model.materials[0].shader, "viewspaceMVP");
    SetShaderValueMatrix(model.materials[0].shader, mvpLoc, mvp);

    // Draw the weapon
    DrawModel(model, position, 1.0, WHITE);
}
```

The params in the function are manipulated by gui sliders.

**Comment by ssansovich** [comment]
Also, the model is just a cube so it shouldn't be that.

`Model weapon_model = LoadModelFromMesh(GenMeshCube(0.1f, 0.4f, 0.1f));`

And I'm pretty sure the shader is correctly accepting the uniform.

**Comment by ssansovich** [comment]
I think I'm making progress.

If I call the draw function inside `BeginShaderMode(shader)` instead of BeginMode3D(camera) then the crazy sheering stops.

My code is now 
```
void draw_weapon(Model model, Camera *camera, float offset_x, float offset_y, float offset_z, float rot_x, float rot_y, float rot_z) {
    Matrix mat_tran = MatrixTranslate(offset_x, offset_y, offset_z);

    Matrix matRotX = MatrixRotateX(rot_x);
    Matrix matRotY = MatrixRotateY(rot_y);
    Matrix matRotZ = MatrixRotateZ(rot_z);
    Matrix mat_rot = MatrixMultiply(MatrixMultiply(matRotZ, matRotY), matRotX);

    Matrix model_matrix = MatrixMultiply(mat_rot, mat_tran);

    // View
    Matrix view = MatrixIdentity();

    // Projection
    Matrix projection = MatrixPerspective(camera->fovy*DEG2RAD, 1600.0f/900.0f, 0.0f, 1000.0f);
    // projection = MatrixIdentity();

    // Compute the model-view-projection matrix
    Matrix mvp = MatrixMultiply(projection, MatrixMultiply(view, model_matrix));

    int mvpLoc = GetShaderLocation(model.materials[0].shader, "viewspaceMVP");
    SetShaderValueMatrix(model.materials[0].shader, mvpLoc, mvp);

    // Draw the gun
    DrawModel(model, Vector3Zero(), 1.0, WHITE); // Position vec isn't used since mvp uniform is used instead
}
```

However, if I use the perspective matrix, the weapon doesn't show up at least as far as I can tell while fiddling with the offsets to make sure its not a culling issue.

If I use an identity matrix for the perspective, everything is working as intended EXCEPT that the model is stretched. I'm assuming this is an issue with the aspect ratio.

Anyone know how to solve this?

**Reply by taylordanger** [reply]
hmmm maybe if you get the screen size it could fix this?


void draw_weapon(Model model, Camera *camera, float offset_x, float offset_y, float offset_z, float rot_x, float rot_y, float rot_z) {
    Matrix mat_tran = MatrixTranslate(offset_x, offset_y, offset_z);
    Matrix matRotX = MatrixRotateX(rot_x);
    Matrix matRotY = MatrixRotateY(rot_y);
    Matrix matRotZ = MatrixRotateZ(rot_z);
    Matrix mat_rot = MatrixMultiply(MatrixMultiply(matRotZ, matRotY), matRotX);

    Matrix model_matrix = MatrixMultiply(mat_rot, mat_tran);

    // View (still identity since you're rendering relative to the camera)
    Matrix view = MatrixIdentity();

    // Projection with corrected aspect ratio and near plane
    float aspect = (float)GetScreenWidth() / (float)GetScreenHeight();
    Matrix projection = MatrixPerspective(camera->fovy*DEG2RAD, aspect, 0.01f, 1000.0f);

    // Compute the model-view-projection matrix
    Matrix mvp = MatrixMultiply(projection, MatrixMultiply(view, model_matrix));

    int mvpLoc = GetShaderLocation(model.materials[0].shader, "viewspaceMVP");
    SetShaderValueMatrix(model.materials[0].shader, mvpLoc, mvp);

    // Draw the gun
    DrawModel(model, Vector3Zero(), 1.0, WHITE);
}


im probably wrong tho
